packages/common-all/src/dnode.ts

```diff
@@ -637,6 +637,123 @@ export class NoteUtils {
     return now;
   }
 
+  /** Returns true when the note has reference links (Eg. ![[ref-link]]); false otherwise. */
+  static hasRefLinks(note: NoteProps) {
+    return note.links.some((link) => link.type === "ref");
+  }
+
+  /**
+   * Retrieve the latest update time of the preview note tree.
+   *
+   * Preview note tree includes links whose content is rendered in the rootNote preview,
+   * particularly the reference links (![[ref-link-example]]). */
+  static getLatestUpdateTimeOfPreviewNoteTree({
+    rootNote,
+    notes,
+  }: {
+    rootNote: NoteProps;
+    notes: NotePropsDict;
+  }) {
+    // If the note does not have reference links than the last updated time of
+    // the preview tree is the last updated time of the note itself. Hence, we
+    // can avoid all heavier logic of getting ready to traverse notes.
+    if (!this.hasRefLinks(rootNote)) {
+      return rootNote.updated;
+    }
+
+    // When there is a vault specified in the link we want to respect that
+    // specification, otherwise we will map by just the file name.
+    // Map by lowercase fname.
+    const mapByFName: { [fname: string]: NoteProps } = {};
+    // Map vault fspath to map which maps by lowercase fname.
+    const mapWithVaultByFName: {
```

Comment:
> This seems to have fixed hover previews not updating, but Preview V2 still doesn't update when referenced notes are changed. Was that meant to be fixed with this PR, or is there additional work needed for that?

The preview is updating on my side. The change is in the server so the test workspace need to be running against the local api-server. Here is a loom video of test scenario with preview v2: https://www.loom.com/share/7336b06b3a3c46f8a2a1252ea08bcec9

> Instead of two maps, could use a single map that maps fnames to a list of notes. 

Sure that makes the code easier on the eyes, refactored.

Reply:
I'm also not seeing the reference updating. See https://www.loom.com/share/95a4948c1186420aabfb2e54986fe72a

Reply:
`test-workspace` was very useful, found two things one was related to getting the vault name correctly and the second one was related to broken links. 

Below are the tests with `test-workspace`

### XVault test:
https://www.loom.com/share/e54dd2b793d04c59a54256bfd9d234ec

### Regular header 
https://www.loom.com/share/970c95a5548343f1895cec5995431800

### Block references
https://www.loom.com/share/da94c5e36fb84d54985dca43eacacc43